modulus/models/module.py

```diff
@@ -65,55 +105,256 @@ def debug(self):
+                self.inner_model = torch_model_class(*args, **kwargs)
+
+            def forward(self, x):
+                return self.inner_model(x)
+
+        # Get the argument names and default values of the PyTorch model's init method
+        init_argspec = inspect.getfullargspec(torch_model_class.__init__)
+        model_argnames = init_argspec.args[1:]  # Exclude 'self'
+        model_defaults = init_argspec.defaults or []
+        defaults_dict = dict(
+            zip(model_argnames[-len(model_defaults) :], model_defaults)
+        )
+
+        # Define the signature of new init
+        params = [inspect.Parameter("self", inspect.Parameter.POSITIONAL_OR_KEYWORD)]
+        params += [
+            inspect.Parameter(
+                argname,
+                inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                default=defaults_dict.get(argname, inspect.Parameter.empty),
+            )
+            for argname in model_argnames
+        ]
+        init_signature = inspect.Signature(params)
+
+        # Replace ModulusModel.__init__ signature with new init signature
+        ModulusModel.__init__.__signature__ = init_signature
+
+        # Generate a unique name for the created class
+        new_class_name = f"{torch_model_class.__name__}ModulusModel"
+        ModulusModel.__name__ = new_class_name
+
+        # Add this class to the dict of dynamically created classes
+        ModulusModel.register(ModulusModel, new_class_name)
+
+        return ModulusModel
+
+    @classmethod
+    def register(cls, model, name: Union[str, None] = None):
+        """
+        Registers a model under a specific name.
+
+        Parameters
+        ----------
+        model
+            The model to be registered.
+        name : Union[str,None], optional
+            The name to register the model under. If none is provided, the model's name
+        """
+
+        # If no name provided, use the model's name
+        if name is None:
+            name = model.__name__
+
+        # Check if name already in use
+        if name in Module._dynamically_created_classes:
+            raise ValueError(f"Name {name} already in use")
+
+        # Add this class to the dict of dynamically created classes
+        Module._dynamically_created_classes[name] = model
+
+    @classmethod
+    def factory(cls, name: str):
+        """
+        Returns a registered model given its name.
+
+        Parameters
+        ----------
+        name : str
+            The name of the registered model.
+
+        Returns
+        -------
+        model
+            The registered model.
+
+        Raises
+        ------
+        KeyError
+            If no model is registered under the provided name.
+        """
+
+        if name in Module._dynamically_created_classes:
+            return Module._dynamically_created_classes[name]
+
+        group = "modulus.models"
+        try:
+            entry_point = entrypoints.get_group_named(group)[name]
```

Comment:
Sorry Oliver, a little confused here. This is entry point in the setup.py? how would this get updated with discovered entry points from other python packages? (Say I pip install my custom NN model with a Modulus entry point)

For example FSspec has this function that will auto register entry points from existing python packages that have been installed. 

https://github.com/fsspec/filesystem_spec/blob/master/fsspec/__init__.py#L40

Did you implement something like this?

Reply:
Outdated, updated in refactor. Resolving.